### PR TITLE
Kernel: Make the construction of VMObject and its child classes OOM-aware

### PR DESCRIPTION
**Pull request by creator1creeper1**
This PR forces the child classes of VMObject to allocate the resources it requires (a `FixedArray<RefPtr<PhysicalPage>>`) for it and pass them to it in its constructor. This resource allocation is done fallibly in their factory functions.

I know that the commits look like they should be squashed together, but I thought it would be more understandable for a reader of the commit history to have incremental atomic improvements in this way rather than one big commit.

The FixedArray changes are based on #11805.

This is work towards #6369.

### CODE REVIEW

File: Kernel/Memory/AnonymousVMObject.cpp [hunk 1]
```diff
@@ -43,7 +43,8 @@ ErrorOr<NonnullRefPtr<VMObject>> AnonymousVMObject::try_clone()
     // one and this one, as well as the clone have sufficient resources
     // to cow all pages as needed
     auto new_shared_committed_cow_pages = TRY(adopt_nonnull_ref_or_enomem(new (nothrow) SharedCommittedCowPages(move(committed_pages))));
-    auto clone = TRY(adopt_nonnull_ref_or_enomem(new (nothrow) AnonymousVMObject(*this, *new_shared_committed_cow_pages)));
+    FixedArray<RefPtr<PhysicalPage>> new_physical_pages = TRY(this->try_clone_physical_pages());
```
Comment (bgianfo) [inline Kernel/Memory/AnonymousVMObject.cpp:46]:
```suggestion
    auto new_physical_pages = TRY(this->try_clone_physical_pages());
```

Comment (bgianfo) [reply Kernel/Memory/AnonymousVMObject.cpp:46]: Here and everywhere else…

**Review by bgianfo — CHANGES_REQUESTED** [review]
Lots of places where we should be using auto instead of explicit types.

File: AK/FixedArray.h [hunk 2]
```diff
@@ -36,6 +36,42 @@ class FixedArray {
...
+        for (size_t i = 0; i < N; ++i)
+            new (&elements[i]) T(move(array[i]));
+        return FixedArray<T>(N, elements);
+    }
+
+    template<size_t N>
+    static FixedArray<T> must_create_but_fixme_should_propagate_errors(T(&&array)[N])
+    {
+        return MUST(try_create(move(array)));
+    }
```
Comment (IdanHo) [inline AK/FixedArray.h:60]: Where is this used? We usually don't add dead code.

Comment (creator1creeper1) [reply AK/FixedArray.h:60]: This is part of #11805 where we already had a small discussion about it. To quote my argument from there:
> The reason I think these APIs are useful is that it extremely lowers the barrier to convert a user to `FixedArray` from some other class like `Vector` that's not OOM-aware. If you had previously created the latter in a constructor, you can now simply use the `must_foo_but_fixme_should_propagate_errors()` to create the former. But if we force every new user to use error propagation, they will need to restructure a lot more code in order to make the switch.

**Comment by creator1creeper1** [comment]
Rebased on master and reworded commits.

**Review by IdanHo — APPROVED** [review]
Cool, thanks

**Pull request merged**